### PR TITLE
[BACKEND-659] Two cloud worker select strategy

### PR DESCRIPTION
**Pull request by leventov**
cc @dgolitsyn

### CODE REVIEW

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ScalingStats.java [hunk 1]
```diff
@@ -65,6 +65,15 @@ public ScalingStats(int capacity)
     }
   }
 
+  public ScalingStats(ScalingStats stats1, ScalingStats stats2)
+  {
+    this(stats1.recentEvents.size() + stats2.recentEvents.size());
+    synchronized (lock) {
```
Comment (egor-ryashin) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ScalingStats.java]: Redundant, we cannot create a single object with 2 threads.
Also, careful, we need somehow to get sizes after a toList() call. Here would be better a constructor like
`new ScalingStats(stats1.toList(), stats2.toList())`

File: indexing-service/src/main/java/io/druid/indexing/overlord/TasksAndWorkers.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+package io.druid.indexing.overlord;
+
+import com.google.common.base.Predicate;
+import io.druid.indexing.common.task.Task;
+import io.druid.indexing.overlord.config.WorkerTaskRunnerConfig;
+import io.druid.indexing.worker.Worker;
+
+import java.util.Collection;
+
+public interface TasksAndWorkers
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TasksAndWorkers.java]: Its hard to track the changes here, does this modify anything beyond what `WorkerTaskRunner` had before the change?

Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TasksAndWorkers.java]: I see `getPendingTasks` here now as well, which is already under `TaskRunner`. Can you explain a little more about this? it seems like it should be able to be refactored in a way that doesn't cause collisions between the two interfaces.

Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TasksAndWorkers.java]: Basically, I don't quite see what splitting into two interfaces instead of an extended interface accomplishes.

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedWorkerProvisioningStrategy.java [hunk 3]
```diff
@@ -268,6 +273,17 @@ private int getWorkersNeededToAssignTasks(
       return need;
     }
 
+    private ImmutableWorkerInfo createDummyWorker(String host, int capacity, String version)
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedWorkerProvisioningStrategy.java]: suggest moving this down in the class to where it was for easier review.

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ScalingStats.java (hunk 1 above)
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ScalingStats.java]: It can matter if the object is being accessed by multiple threads. Only final fields are guaranteed memory consistent after the constructor. Any objects in a list in the final field (for example) are not guaranteed memory consistent.

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  public TasksAndWorkersFilteredByIp(WorkerTaskRunner delegate, String ipPrefix, Predicate<Task> taskPredicate)
+  {
+    this.delegate = delegate;
+    this.ipPrefix = ipPrefix;
+    this.taskPredicate = taskPredicate;
+  }
+
+  private Collection<Worker> filter(Collection<Worker> workers)
+  {
+    Collection<Worker> filtered = new ArrayList<>(workers.size());
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java]: this can be made quite functional with

```java
return ImmutableList.copyOf(Collections2.filter(workers, ipPredicate))
```

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  }
+
+  @Override
+  public Collection<ImmutableWorkerInfo> getWorkers()
+  {
+    Collection<ImmutableWorkerInfo> workers = delegate.getWorkers();
+    Collection<ImmutableWorkerInfo> filtered = new ArrayList<>(workers.size());
+    for (ImmutableWorkerInfo workerInfo : workers) {
+      if (workerInfo.getWorker().getIp().startsWith(ipPrefix)) {
+        filtered.add(workerInfo);
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java]: Also suggest making this more functional with `Collections2`

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  @Override
+  public Collection<Worker> getLazyWorkers()
+  {
+    return filter(delegate.getLazyWorkers());
+  }
+
+  @Override
+  public Collection<Worker> markWorkersLazy(final Predicate<ImmutableWorkerInfo> isLazyWorker, int maxWorkers)
+  {
+    Predicate<ImmutableWorkerInfo> isLazyWorkerWithProperIp = new Predicate<ImmutableWorkerInfo>()
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java]: suggest using `com.google.common.base.Predicates#and(java.lang.Iterable<? extends com.google.common.base.Predicate<? super T>>)` or similar

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    return delegate.getConfig();
+  }
+
+  @Override
+  public Collection<Task> getPendingTaskPayloads()
+  {
+    ImmutableList.Builder<Task> filtered = ImmutableList.builder();
+    for (Task task : delegate.getPendingTaskPayloads()) {
+      if (taskPredicate.apply(task)) {
+        filtered.add(task);
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java]: suggest guava handy `Collections2` here as well

File: indexing-service/src/main/java/io/druid/indexing/overlord/WorkerTaskRunner.java [hunk 8]
```diff
@@ -18,37 +18,7 @@
...
 
-import com.google.common.base.Predicate;
-import io.druid.indexing.common.task.Task;
-import io.druid.indexing.overlord.config.WorkerTaskRunnerConfig;
-import io.druid.indexing.worker.Worker;
 
-import java.util.Collection;
-
-public interface WorkerTaskRunner extends TaskRunner
+public interface WorkerTaskRunner extends TaskRunner, TasksAndWorkers
```
Comment (clintropolis) [inline indexing-service/src/main/java/io/druid/indexing/overlord/WorkerTaskRunner.java]: I think I'm missing what is gained by splitting `TasksAndWorkers` out from `WorkerTaskRunner` which is essentially empty now?

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+      }
+    }
+    return filtered.build();
+  }
+
+  @Override
+  public Collection<? extends TaskRunnerWorkItem> getPendingTasks()
+  {
+    Collection<? extends TaskRunnerWorkItem> pendingTasks = delegate.getPendingTasks();
+    Set<String> pendingTaskIds = new HashSet<>();
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java]: This one could be done with a combination of `com.google.common.collect.Collections2#transform` and filter, but I'm not sure if it would be much cleaner.

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TwoCloudWorkerProvisioningStrategy.java [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+import io.druid.indexing.overlord.setup.WorkerBehaviorConfig;
+
+import javax.annotation.Nullable;
+import java.util.concurrent.ScheduledExecutorService;
+
+import static io.druid.indexing.common.task.TaskLabels.getTaskLabel;
+
+public class TwoCloudWorkerProvisioningStrategy extends AbstractWorkerProvisioningStrategy
+{
+  public static final Supplier<ScheduledExecutorService> EXEC_FACTORY = new Supplier<ScheduledExecutorService>()
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TwoCloudWorkerProvisioningStrategy.java]: Can this just be `Execs.scheduledSingleThreaded("TwoCloudWorkerProvisioningStrategy-manager--%d")`?

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TwoCloudWorkerProvisioningStrategy.java [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    );
+    final Provisioner provisioner2 = provisioningStrategy2.makeProvisioner(
+        new TasksAndWorkersFilteredByIp((WorkerTaskRunner) runner, ipPrefix2, taskPredicate2)
+    );
+    return new Provisioner()
+    {
+      @Override
+      public boolean doTerminate()
+      {
+        return provisioner1.doTerminate() || provisioner2.doTerminate();
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TwoCloudWorkerProvisioningStrategy.java]: the second one won't fire if the first one returns true, if I recall. This should fire both regardless of either's return value

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ScalingStats.java [hunk 12]
```diff
@@ -65,6 +65,15 @@ public ScalingStats(int capacity)
     }
   }
 
+  public ScalingStats(ScalingStats stats1, ScalingStats stats2)
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ScalingStats.java]: IMHO it would be better to have a `addAll(Collection<? extends ScalingEvent> events)` kind of method instead of a constructor that only takes 2.

File: indexing-service/src/main/java/io/druid/indexing/overlord/TwoCloudWorkerSelectStrategy.java [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+    this.twoCloudConfig = twoCloudConfig;
+  }
+
+  @Override
+  public Optional<ImmutableWorkerInfo> findWorkerForTask(
+      WorkerTaskRunnerConfig config, ImmutableMap<String, ImmutableWorkerInfo> zkWorkers, Task task
+  )
+  {
+    String taskLabel = getTaskLabel(task);
+    String ipFilter = twoCloudConfig.getIpFilter(taskLabel);
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TwoCloudWorkerSelectStrategy.java]: Eventually this kind of logic could be extended to an arbitrary number of cloud providers in a chain where you waterfall through the predicates to see who accepts it.

File: indexing-service/src/main/java/io/druid/indexing/overlord/WorkerTaskRunner.java (hunk 8 above)
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/WorkerTaskRunner.java]: @clintropolis `WorkerTaskRunner` also extends `TaskRunner`, which contains a lot of methods which I didn't want to implement in `TasksAndWorkersFilteredByIp`.

File: services/src/main/java/io/druid/cli/CliOverlord.java [hunk 14]
```diff
@@ -205,6 +207,7 @@ private void configureRunners(Binder binder)
             binder.bind(RemoteTaskRunnerFactory.class).in(LazySingleton.class);
 
             JacksonConfigProvider.bind(binder, WorkerBehaviorConfig.CONFIG_KEY, WorkerBehaviorConfig.class, null);
+            JacksonConfigProvider.bind(binder, WorkerBehaviorConfig.CONFIG_KEY + "selectStrategy.twoCloudConfig", TwoCloudConfig.class, null);
```
Comment (leventov) [inline services/src/main/java/io/druid/cli/CliOverlord.java]: Added

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TwoCloudWorkerProvisioningStrategy.java [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+      }
+
+      @Override
+      public boolean doProvision()
+      {
+        return provisioner1.doProvision() || provisioner2.doProvision();
```
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TwoCloudWorkerProvisioningStrategy.java]: Good catch, fixed

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ScalingStats.java (hunk 1 above)
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ScalingStats.java]: Refactored in `ScalingStats.addAll()` and no new constructor.

@egor-ryashin capacity=0 is handled in the existing constructor of `ScalingStats`

File: indexing-service/src/main/java/io/druid/indexing/overlord/TwoCloudWorkerSelectStrategy.java [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+import io.druid.indexing.overlord.config.WorkerTaskRunnerConfig;
+import io.druid.indexing.overlord.setup.TwoCloudConfig;
+import io.druid.indexing.overlord.setup.WorkerBehaviorConfig;
+import io.druid.indexing.overlord.setup.WorkerSelectStrategy;
+
+import java.util.Map;
+
+import static io.druid.indexing.common.task.TaskLabels.getTaskLabel;
+
+public class TwoCloudWorkerSelectStrategy implements WorkerSelectStrategy
```
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TwoCloudWorkerSelectStrategy.java]: Can move after review

File: indexing-service/src/main/java/io/druid/indexing/overlord/TwoCloudWorkerSelectStrategy.java (hunk 13 above)
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TwoCloudWorkerSelectStrategy.java]: I prefer to keep things simple i. e. the way they are now, until more generality is actually needed

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ScalingStats.java (hunk 12 above)
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ScalingStats.java]: Refactored

File: indexing-service/src/main/java/io/druid/indexing/overlord/TasksAndWorkers.java (hunk 2 above)
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TasksAndWorkers.java]: I didn't want to make `TasksAndWorkersFilteredByIp` to implement `WorkerTaskRunner` because there are a lot of methods in `TaskRunner` which don't make sense for the interface, needed in Provisioners. And not clear how to implement them.

I agree that something is wrong here and some refactoring is needed, but it would affect a lot of unrelated code, so I don't want to do this until a PR is done into Druid upstream.

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedWorkerProvisioningStrategy.java (hunk 3 above)
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedWorkerProvisioningStrategy.java]: It changed the nesting level (moved to an inner class)

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java (hunk 4 above)
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java]: Couldn't be done without anonymous lambda, that is much more cumbersome before Java 8 than the present code

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java (hunk 5 above)
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java]: Fixed

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java (hunk 6 above)
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java]: Fixed

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java (hunk 7 above)
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java]: Fixed

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java (hunk 9 above)
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java]: Before Java 8 it won't be cleaner, it will be only more cumbersome

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TwoCloudWorkerProvisioningStrategy.java (hunk 10 above)
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TwoCloudWorkerProvisioningStrategy.java]: Fixed

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TwoCloudWorkerProvisioningStrategy.java (hunk 11 above)
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TwoCloudWorkerProvisioningStrategy.java]: Good catch, fixed

**Comment by leventov** [comment]
Thanks @egor-ryashin @clintropolis @drcrallen

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TwoCloudWorkerProvisioningStrategy.java [hunk 17]
```diff
@@ -121,19 +131,30 @@ Provisioner makeProvisioner(TasksAndWorkers runner)
...
+        return provisioned1 || provisioned2;
       }
 
       @Override
       public ScalingStats getStats()
       {
-        return new ScalingStats(provisioner1.getStats(), provisioner2.getStats());
+        ScalingStats stats1 = provisioner1.getStats();
+        ScalingStats stats2 = provisioner2.getStats();
+        ScalingStats stats = new ScalingStats(stats1.size() + stats2.size());
```
Comment (egor-ryashin) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TwoCloudWorkerProvisioningStrategy.java]: We could lose some initial events, for example, we call stats1.size() (when it is 0) than **before** we call stats1.toList() other thread can quickly add to stats1, but we already set max size 0 for a new queue, and we lose those events.

Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TwoCloudWorkerProvisioningStrategy.java]: @egor-ryashin I see now, thanks. Fixed

**Comment by leventov** [comment]
@clintropolis @drcrallen @egor-ryashin @dgolitsyn I've added support of different dynamic configs to `TwoCloudWorkerProvisioningStrategy`, please review

File: common/src/main/java/io/druid/audit/AuditEntry.java [hunk 18]
```diff
@@ -91,6 +90,12 @@ public DateTime getAuditTime()
     return auditTime;
   }
 
+  @Override
+  public int compareTo(AuditEntry otherEntry)
```
Comment (drcrallen) [inline common/src/main/java/io/druid/audit/AuditEntry.java]: This has an inherent race condition for timers that only have resolution millisecond. Why is this needed?

Comment (drcrallen) [inline common/src/main/java/io/druid/audit/AuditEntry.java]: Since this seems to only be used to sort for returning results, can you imply have a comparator parameter to the sort instead of making the compare intrinsic to the class?

Comment (leventov) [inline common/src/main/java/io/druid/audit/AuditEntry.java]: Changed

**Comment by leventov** [comment]
Closing this PR because review is done. It's more convenient to review ongoing changes separately.

**Comment by leventov** [comment]
Please don't delete branch yet, it is not yet merged into `specialize-topn-queries`
